src/compiler/checker.ts

```diff
@@ -8884,6 +8896,17 @@ namespace ts {
         function getResolvedMembersOrExportsOfSymbol(symbol: Symbol, resolutionKind: MembersOrExportsResolutionKind): UnderscoreEscapedMap<Symbol> {
             const links = getSymbolLinks(symbol);
             if (!links[resolutionKind]) {
+                // We _must_ resolve any possible commonjs export assignments _before_ merging in late bound members, as cjs export assignments may cause
+                // things to be merged (?!?) into this symbol; moreover, we _can't_ resolve those export assignments if we're already resolving the containing
+                // module (as then we'll issue a circularity error)

```

Comment:
The reason I haven’t approved this PR is because I don’t really understand this (or really, why changes in alias resolution were involved generally), so I was hoping someone else would. But since nobody else has come by, if you want to expound on this, I’ll take a closer look.

Reply:
I agree; it seems like this is needed to support checking, not necessarily declaration emit, since it skips late-bound symbols anyway.

Reply:
The checking changed because the declaration emit made it evident it was wrong. We were using `symbol.exports` before; this was missing late bound members, so we swapped to using `getExportsOfSymbol` in this PR; however using that made it obvious that the resolved exports were being cached without any cjs merged exports in them! (So symbol.exports had members that were never reflected in the resolved exports!) That was because we were not reliably doing the cjs merge before caching the resolved exports; this new check guarantees that we try to do the cjs merge _before_ caching resolved exports, every time.